usaspending_api/recipient/v2/views.py

```diff
@@ -0,0 +1,117 @@
+import logging
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+from collections import OrderedDict
+
+from rest_framework.response import Response
+from usaspending_api.common.views import APIDocumentationView
+
+from django.db.models import Sum
+
+from usaspending_api.awards.models_matviews import UniversalTransactionView
+from usaspending_api.awards.v2.filters.filter_helpers import sum_transaction_amount
+# from usaspending_api.awards.v2.filters.sub_award import subaward_filter
+from usaspending_api.awards.v2.filters.transaction import transaction_filter
+from usaspending_api.awards.v2.filters.matview_filters import universal_transaction_matview_filter
+from usaspending_api.common.exceptions import InvalidParameterException
+from usaspending_api.common.helpers import table_exists, generate_fiscal_year
+from usaspending_api.recipient.models import StateData
+
+logger = logging.getLogger(__name__)
+
+
+class StateMetaDataViewSet(APIDocumentationView):
+
+    def get_state_data(self, state_data_results, field, year=None):
+        """Finds which earliest or latest state data to use based on the year and what data is available"""
+        state_data = OrderedDict(sorted([(str(state_data['year']), state_data) for state_data in state_data_results
+                                         if state_data[field]], key=lambda pair: pair[0]))
+        earliest = list(state_data.keys())[0]
+        latest = list(state_data.keys())[-1]
+        if year and year.isdigit() and year < earliest:
+            return state_data[earliest]
+        elif year and year.isdigit() and earliest <= year <= latest:
```

Comment:
this check is repeated twice, could we check this once up front and raise the appropriate exception? then subsequent calls only check the relevant cases.

Input validation like this would also be covered by TinyShield.

Reply:
Just looked over TinyShield and it doesn't currently support multiple types for a field. Example: the year param can be a number but it can also be `'latest'` and `'all'`